File: qif4j/src/main/java/com/github/squirrelgrip/qif/QifCashTransaction.java
package com.github.squirrelgrip.qif;

import java.math.BigDecimal;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class QifCashTransaction extends QifTransaction {

	private List<String> address = new ArrayList<String>();
	private String category;
	private String clearedStatus;
	private QifSplitTransaction currentSplitTransaction;
	private String payee;
	private String reference;
	private List<QifSplitTransaction> splitTransactions = new ArrayList<QifSplitTransaction>();

	public void addAddress(String address) {
		this.address.add(address);
	}

	private void checkSplitTransaction() {
		if (currentSplitTransaction == null) {
			createSplitTransaction();
		}
	}
	
	private void createSplitTransaction() {
		currentSplitTransaction = new QifSplitTransaction(this);
		splitTransactions.add(currentSplitTransaction);
	}
	
	public List<String> getAddress() {
		return address;
	}
	
	public String getCategory() {
		return category;
	}
	
	public String getClearedStatus() {
		return clearedStatus;
	}

	public QifSplitTransaction getCurrentSplitTransaction() {
		return currentSplitTransaction;
	}

	public String getPayee() {
		return payee;
	}

	public String getReference() {
		return reference;
	}

	public List<QifSplitTransaction> getSplits() {
		return splitTransactions;
	}

	public List<QifSplitTransaction> getSplitTransactions() {
		return splitTransactions;
	}

	public void setCategory(String category) {
		this.category = category;
	}

	public void setClearedStatus(String clearedStatus) {
		this.clearedStatus = clearedStatus;
	}

	public void setPayee(String payee) {
		this.payee = payee;
	}

	public void setReference(String reference) {
		this.reference = reference;
	}
	
	public void setSplitAmount(String splitAmount) {
		checkSplitTransaction();
		if(currentSplitTransaction.getAmount() != null) {
			createSplitTransaction();
		}
		currentSplitTransaction.setAmount(new BigDecimal(splitAmount));
	}
	
	public void setSplitCategory(String splitCategory) {
		checkSplitTransaction();
		if(currentSplitTransaction.getCategory() != null) {
			createSplitTransaction();
		}
		currentSplitTransaction.setCategory(splitCategory);
	}

	public void setSplitMemo(String splitMemo) {
		checkSplitTransaction();
		if(currentSplitTransaction.getMemo() != null) {
			createSplitTransaction();
		}
		currentSplitTransaction.setMemo(splitMemo);
	}
	
	public String toCsv(String dateFormat) {
		return String.format("%s,%s,%s,%s,%s", getDate().format(DateTimeFormatter.ofPattern(dateFormat)), empty(getCategory()),
		        empty(getPayee()), getTotal(), empty(getMemo()));
	}
	
	private String empty(String value) {
		return value == null ? "" : value;
	}

}

File: qif4k/src/main/kotlin/com/github/squirrelgrip/qif4k/QifCashTransaction.kt
package com.github.squirrelgrip.qif4k

import java.math.BigDecimal
import java.time.format.DateTimeFormatter
import java.util.*

class QifCashTransaction : QifTransaction() {
    private val _address: MutableList<String> = mutableListOf()
    var category: String? = null
    var clearedStatus: String? = null
    private var currentSplitTransaction: QifSplitTransaction? = null
    var payee: String? = null
    var reference: String? = null
    private val splitTransactions: MutableList<QifSplitTransaction> = ArrayList()

    fun addAddress(address: String) {
        this._address.add(address)
    }

    private fun checkSplitTransaction() {
        if (currentSplitTransaction == null) {
            createSplitTransaction()
        }
    }

    private fun createSplitTransaction() {
        currentSplitTransaction = QifSplitTransaction(this)
        splitTransactions.add(currentSplitTransaction!!)
    }

    val address: List<String>
        get() = _address.toList()


    val splits: List<QifSplitTransaction>
        get() = splitTransactions

    val getSplitTransactions: List<QifSplitTransaction>
        get() = splitTransactions


    fun setSplitAmount(splitAmount: String) {
        checkSplitTransaction()
        if (currentSplitTransaction!!.amount != null) {
            createSplitTransaction()
        }
        currentSplitTransaction!!.amount = BigDecimal(splitAmount)
    }

    fun setSplitCategory(splitCategory: String?) {
        checkSplitTransaction()
        if (currentSplitTransaction!!.category != null) {
            createSplitTransaction()
        }
        currentSplitTransaction!!.category = splitCategory
    }

    fun setSplitMemo(splitMemo: String?) {
        checkSplitTransaction()
        if (currentSplitTransaction!!.memo != null) {
            createSplitTransaction()
        }
        currentSplitTransaction!!.memo = splitMemo
    }

    fun toCsv(dateFormat: String?): String {
        return String.format("%s,%s,%s,%s,%s", date?.format(DateTimeFormatter.ofPattern(dateFormat)), empty(category),
                empty(payee), total, empty(memo))
    }

    private fun empty(value: String?): String {
        return value ?: ""
    }
}